codes/go/chapter_hashing/array_hash_map.go

```diff
@@ -0,0 +1,96 @@
+// File: array_hash_map.go
+// Created Time: 2022-12-14
+// Author: msk397 (machangxinq@gmail.com)
+
+package chapter_hashing
+
+import "fmt"
+
+// 键值对 int->String
+type Entry struct {
+	key int
+	val string
+}
+
+// 基于数组简易实现的哈希表
+type ArrayHashMap struct {
+	bucket []Entry
+}
+
+func newArrayHashMap() *ArrayHashMap {
+	// 初始化一个长度为 100 的桶（数组）
+	bucket := make([]Entry, 100)
+	return &ArrayHashMap{bucket: bucket}
+}
+
+// 哈希函数
+func (a *ArrayHashMap) hashFunc(key int) int {
+	index := key % 100
+	return index
+}
+
+// 查询操作
+func (a *ArrayHashMap) get(key int) string {
+	index := a.hashFunc(key)
+	pair := a.bucket[index]
+	if pair.key == 0 {
+		return ""
+	}
+	return pair.val
+}
+
+//  添加操作
+func (a *ArrayHashMap) put(key int, val string) {
+	pair := Entry{key: key, val: val}
+	index := a.hashFunc(key)
+	a.bucket[index] = pair
+}
+
+// 删除操作
+func (a *ArrayHashMap) remove(key int) {
+	index := a.hashFunc(key)
+	// 置为空字符，代表删除
+	a.bucket[index] = Entry{}
+}
+
+// 获取所有键对
+func (a *ArrayHashMap) entrySet() []Entry {
+	var pairs []Entry
+	for _, pair := range a.bucket {
+		if pair.key != 0 {
+			pairs = append(pairs, pair)
+		}
+	}
+	return pairs
+}
+
+// 获取所有键
+func (a *ArrayHashMap) keySet() []int {
+	var keys []int
+	for _, pair := range a.bucket {
+		if pair.key != 0 {
```

Comment:
What will happen if there is a pair whose key = 0 ?

Reply:
I thought int type can't be nil, maybe I'm wrong, I'll try your suggestion

Reply:
Could `pair` be nil?